File: engine/__tests__/predictedAeProfile.test.ts
import { predictAeProfile, likelihoodLabel } from '../predictedAeProfile';
import { getDrug } from '../switchingEngine';

describe('predictAeProfile', () => {
  test('aripiprazole shows akathisia as high (drug profile flag)', () => {
    const profile = predictAeProfile(getDrug('aripiprazole')!);
    const eps = profile.predictions.find((p) => p.ae.id === 'eps_akathisia');
    expect(eps).toBeDefined();
    // Aripiprazole is in causedBy[] for eps_akathisia → high likelihood
    expect(['high', 'moderate']).toContain(eps!.likelihood);
  });

  test('olanzapine shows weight gain as high', () => {
    const profile = predictAeProfile(getDrug('olanzapine')!);
    const wt = profile.predictions.find((p) => p.ae.id === 'weight_gain');
    expect(wt?.likelihood).toBe('high');
  });

  test('aripiprazole shows weight gain as low when switching from olanzapine', () => {
    const profile = predictAeProfile(getDrug('aripiprazole')!, getDrug('olanzapine')!);
    const wt = profile.predictions.find((p) => p.ae.id === 'weight_gain');
    expect(wt?.likelihood).toBe('lower-than-current');
  });

  test('mirtazapine shows sedation high (per drug profile)', () => {
    const profile = predictAeProfile(getDrug('mirtazapine')!);
    const sed = profile.predictions.find((p) => p.ae.id === 'sedation');
    expect(sed).toBeDefined();
    expect(['high', 'moderate']).toContain(sed!.likelihood);
  });

  test('predictions sorted by likelihood (high first)', () => {
    const profile = predictAeProfile(getDrug('clozapine')!);
    const ranks = profile.predictions.map((p) => {
      switch (p.likelihood) {
        case 'high': return 4;
        case 'moderate': return 3;
        case 'low': return 2;
        case 'lower-than-current': return 1;
        case 'unknown': return 0;
      }
    });
    for (let i = 1; i < ranks.length; i++) {
      expect(ranks[i]).toBeLessThanOrEqual(ranks[i - 1]);
    }
  });

  test('every prediction includes a non-empty reason', () => {
    const profile = predictAeProfile(getDrug('quetiapine')!);
    for (const p of profile.predictions) {
      expect(p.reason.length).toBeGreaterThan(0);
    }
  });

  test('likelihoodLabel returns user-facing strings', () => {
    expect(likelihoodLabel('high')).toMatch(/high/i);
    expect(likelihoodLabel('lower-than-current')).toMatch(/lower/i);
  });
});

File: flutter_app/test/engine/predicted_ae_profile_test.dart
// Tests for the Dart predicted_ae_profile port.
// Mirrors engine/__tests__/predictedAeProfile.test.ts, but uses inline
// Drug fixtures rather than getDrug() (which we haven't ported yet).

import 'package:flutter_test/flutter_test.dart';
import 'package:psychswitch/src/engine/predicted_ae_profile.dart';
import 'package:psychswitch/src/engine/types/drug.dart';
import 'package:psychswitch/src/engine/types/enums.dart';

const _emptyMetabolite = ActiveMetabolite(
  name: null,
  halfLifeHours: null,
  clinicallySignificant: false,
);
const _emptyCyp = CypInteractions(
  substrateOf: <String>[],
  inhibitorOf: <String>[],
  switchingRelevance: '',
);
const _emptyDosing = Dosing(
  startingDoseMg: 0,
  typicalTargetRangeMg: <double>[],
  maxDoseMg: 0,
  increments: <double>[],
  formulationsAvailableMy: <String>[],
);

Drug _drug({
  required String id,
  RiskLevel? sedation,
  RiskLevel? epsRisk,
  RiskLevel? prolactinRisk,
  RiskLevel? qtcRisk,
  MetabolicRisk? metabolicRisk,
  DiscontinuationSyndromeRisk? discontinuationSyndromeRisk,
}) =>
    Drug(
      id: id,
      genericName: id,
      drugClass: 'antidepressant',
      malaysianBrandNames: const <String>[],
      halfLife: const HalfLife(meanHours: 24, rangeHours: <double>[]),
      activeMetabolite: _emptyMetabolite,
      cypInteractions: _emptyCyp,
      sedation: sedation,
      epsRisk: epsRisk,
      prolactinRisk: prolactinRisk,
      qtcRisk: qtcRisk,
      metabolicRisk: metabolicRisk,
      discontinuationSyndromeRisk: discontinuationSyndromeRisk,
      dosing: _emptyDosing,
      formulationNotes: '',
      citations: const <String>[],
      lastReviewedISO: '2026-04-01',
      reviewedBy: 'Test',
    );

void main() {
  group('predictAeProfile', () {
    test(
      'aripiprazole shows akathisia as high (reverse-lookup wins over '
      'low drug field)',
      () {
        final profile = predictAeProfile(
          _drug(
            id: 'aripiprazole',
            epsRisk: RiskLevel.low, // drug field says low (parkinsonism)
          ),
        );
        final eps = profile.predictions.firstWhere(
          (p) => p.ae.id == 'eps_akathisia',
        );
        expect(
          <AeLikelihood>[AeLikelihood.high, AeLikelihood.moderate],
          contains(eps.likelihood),
        );
      },
    );

    test('olanzapine shows weight gain as high', () {
      final profile = predictAeProfile(
        _drug(
          id: 'olanzapine',
          metabolicRisk: const MetabolicRisk(
            score: RiskLevel.high,
            notes: '',
          ),
        ),
      );
      final wt = profile.predictions.firstWhere(
        (p) => p.ae.id == 'weight_gain',
      );
      expect(wt.likelihood, equals(AeLikelihood.high));
    });

    test(
      'aripiprazole shows weight gain as lower-than-current when switching '
      'from olanzapine',
      () {
        final profile = predictAeProfile(
          _drug(id: 'aripiprazole'),
          _drug(id: 'olanzapine'),
        );
        final wt = profile.predictions.firstWhere(
          (p) => p.ae.id == 'weight_gain',
        );
        expect(wt.likelihood, equals(AeLikelihood.lowerThanCurrent));
      },
    );

    test('mirtazapine shows sedation high (per drug profile)', () {
      final profile = predictAeProfile(
        _drug(id: 'mirtazapine', sedation: RiskLevel.high),
      );
      final sed = profile.predictions.firstWhere(
        (p) => p.ae.id == 'sedation',
      );
      expect(
        <AeLikelihood>[AeLikelihood.high, AeLikelihood.moderate],
        contains(sed.likelihood),
      );
    });

    test('predictions sorted by likelihood (high first)', () {
      final profile = predictAeProfile(
        _drug(
          id: 'clozapine',
          sedation: RiskLevel.high,
          metabolicRisk: const MetabolicRisk(
            score: RiskLevel.high,
            notes: '',
          ),
        ),
      );
      final ranks = profile.predictions
          .map((p) => switch (p.likelihood) {
                AeLikelihood.high => 4,
                AeLikelihood.moderate => 3,
                AeLikelihood.low => 2,
                AeLikelihood.lowerThanCurrent => 1,
                AeLikelihood.unknown => 0,
              })
          .toList();
      for (var i = 1; i < ranks.length; i++) {
        expect(ranks[i], lessThanOrEqualTo(ranks[i - 1]));
      }
    });

    test('every prediction includes a non-empty reason', () {
      final profile = predictAeProfile(
        _drug(
          id: 'quetiapine',
          sedation: RiskLevel.high,
          metabolicRisk: const MetabolicRisk(
            score: RiskLevel.moderate,
            notes: '',
          ),
        ),
      );
      for (final p in profile.predictions) {
        expect(p.reason, isNotEmpty);
      }
    });

    test('drug not in any AE table returns empty predictions', () {
      final profile = predictAeProfile(_drug(id: 'unknown-drug'));
      expect(profile.predictions, isEmpty);
    });
  });

  group('likelihoodLabel', () {
    test('returns user-facing strings', () {
      expect(
        likelihoodLabel(AeLikelihood.high).toLowerCase(),
        contains('high'),
      );
      expect(
        likelihoodLabel(AeLikelihood.lowerThanCurrent).toLowerCase(),
        contains('lower'),
      );
    });
  });

  group('AeLikelihood jsonValue round-trips', () {
    test('every likelihood parses back', () {
      for (final l in AeLikelihood.values) {
        expect(AeLikelihood.fromJson(l.jsonValue), equals(l));
      }
    });
  });
}
